### PR TITLE
unshare: Add options to identity map the user's subordinate uids and gids

### PR DESCRIPTION
**Pull request by dgibson**
`--map-users=auto`, `--map-groups=auto` and `--map-auto` allow automatically mapping the first block of user or group IDs owned by the effective user from /etc/sub[ug]id to a block starting at user ID 0.  Add options `--map-users=subids`, `--map-groups=subids` and `--map-subids` to perform a mapping of the same IDs, but as an identity mapping instead of to UID 0.

This is useful in similar situations to --map-current-user, but preserves access to the user's subordinate IDs.  That allows nested namespaces to be created with more complex user mappings, either with unshare or with a container runtime such as podman.

Fixes: https://github.com/util-linux/util-linux/issues/3120

### CODE REVIEW

**Comment by dgibson** [comment]
Example usage.  Without mapping subordinate ids you can't run `podman` in an `unshare`d namespace:

```
$ unshare -Uc
$ podman run -it --rm docker.io/library/busybox
ERRO[0000] running `/bin/newuidmap 112828 0 1000 1 1 524288 65536`: newuidmap: write to uid_map failed: Operation not permitted 
ERRO[0000] invalid internal status, try resetting the pause process with "podman system migrate": cannot set up namespace using "/bin/newuidmap": exit status 1 
```

These new options make it easy to establish the necessary mappings
```
$ unshare -Uc --map-subids
$ podman run -it --rm docker.io/library/busybox
/ # 
```

**Comment by karelzak** [comment]
@arachsys and @brauner, do you have any objections or comments on this feature?

**Comment by arachsys** [comment]
This makes a lot of sense to me. --map-auto was a generalisation of --map-root-user to map subids sensibly. The proposed --map-subids is the analogous generalisation of --map-currrent-user to map subids sensibly. I know I've constructed exactly these pass-through mappings by hand before in scripts, transforming /proc/self/[ug]id_map into --map-* arguments.

**Comment by dgibson** [comment]
Btw, I have no objections to changing the names of the options if you have better ideas.

**Comment by karelzak** [comment]
I think the `subids` name is OK. I'll merge it.

**Comment by brauner** [comment]
> @arachsys and @brauner, do you have any objections or comments on this feature?

Seems good to me!

File: sys-utils/unshare.c [hunk 1]
```diff
@@ -990,16 +1001,24 @@ int main(int argc, char *argv[])
 			unshare_flags |= CLONE_NEWUSER;
 			if (!strcmp(optarg, "auto"))
 				insert_map_range(&groupmap,
-					read_subid_range(_PATH_SUBGID, real_euid));
+						 read_subid_range(_PATH_SUBGID, real_euid, 0));
+			else if (!strcmp(optarg, "subids"))
+				insert_map_range(&usermap,
+						 read_subid_range(_PATH_SUBGID, real_euid, 1));
```
Comment (karelzak) [inline sys-utils/unshare.c:1007]: @dgibson coverity-scan reports this "&usermap" as a possible copy & past bug, and it really seems that we want "&groupmap" here. It's OPT_MAPGROUPS code block (--map-groups).  Can you please check the code?

Comment (dgibson) [reply sys-utils/unshare.c:1007]: Yes, it should be `groupmap`.  Embarrasing.  Interestingly it's correct for `--map-subids` but not for `--map-groups=subids`.

You want me to send a fixup patch? Or will you handle it?

Comment (karelzak) [reply sys-utils/unshare.c:1007]: I'll fix it. Thanks.

**Pull request merged**